File: task.ts
/**
 * Task management for Produco - TypeScript version
 * Type-safe task management
 */

import { Task } from './api';

export class TaskManager {
  private tasks: Task[] = [];
  private storage: any;

  constructor(storage: any) {
    this.storage = storage;
    this.loadTasks();
  }

  private loadTasks(): void {
    const data = this.storage.getItem('tasks');
    if (data) {
      this.tasks = JSON.parse(data);
    }
  }

  private saveTasks(): void {
    this.storage.setItem('tasks', JSON.stringify(this.tasks));
  }

  addTask(title: string, description: string = '', priority: 'low' | 'medium' | 'high' = 'medium', dueDate?: string): Task {
    const task: Task = {
      id: Date.now(),
      title,
      description,
      priority,
      completed: false,
      createdAt: new Date().toISOString(),
      dueDate
    };
    this.tasks.push(task);
    this.saveTasks();
    return task;
  }

  completeTask(taskId: number): Task | null {
    const task = this.tasks.find(t => t.id === taskId);
    if (task && !task.completed) {
      task.completed = true;
      task.completedAt = new Date().toISOString();
      this.saveTasks();
      return task;
    }
    return null;
  }

  deleteTask(taskId: number): boolean {
    const index = this.tasks.findIndex(t => t.id === taskId);
    if (index !== -1) {
      this.tasks.splice(index, 1);
      this.saveTasks();
      return true;
    }
    return false;
  }

  getTask(taskId: number): Task | undefined {
    return this.tasks.find(t => t.id === taskId);
  }

  getAllTasks(): Task[] {
    return this.tasks;
  }

  getPendingTasks(): Task[] {
    return this.tasks.filter(t => !t.completed);
  }

  getCompletedTasks(): Task[] {
    return this.tasks.filter(t => t.completed);
  }

  getTasksByPriority(priority: 'low' | 'medium' | 'high'): Task[] {
    return this.tasks.filter(t => t.priority === priority);
  }
}

File: task.rs
/**
 * Task management for Produco - Rust implementation
 * Type-safe task management
 */

use std::collections::HashMap;
use chrono::Utc;

#[derive(Debug, Clone)]
pub struct Task {
    pub id: u32,
    pub title: String,
    pub description: String,
    pub priority: String,
    pub completed: bool,
    pub created_at: String,
    pub due_date: Option<String>,
    pub completed_at: Option<String>,
}

pub struct TaskManager {
    tasks: HashMap<u32, Task>,
    next_id: u32,
}

impl TaskManager {
    pub fn new() -> Self {
        TaskManager {
            tasks: HashMap::new(),
            next_id: 1,
        }
    }
    
    pub fn add_task(&mut self, title: String, description: String, priority: String, due_date: Option<String>) -> Task {
        let task = Task {
            id: self.next_id,
            title,
            description,
            priority,
            completed: false,
            created_at: Utc::now().to_rfc3339(),
            due_date,
            completed_at: None,
        };
        
        self.tasks.insert(self.next_id, task.clone());
        self.next_id += 1;
        
        task
    }
    
    pub fn complete_task(&mut self, task_id: u32) -> Option<&Task> {
        if let Some(task) = self.tasks.get_mut(&task_id) {
            if !task.completed {
                task.completed = true;
                task.completed_at = Some(Utc::now().to_rfc3339());
            }
            return Some(task);
        }
        None
    }
    
    pub fn delete_task(&mut self, task_id: u32) -> bool {
        self.tasks.remove(&task_id).is_some()
    }
    
    pub fn get_task(&self, task_id: u32) -> Option<&Task> {
        self.tasks.get(&task_id)
    }
    
    pub fn get_all_tasks(&self) -> Vec<&Task> {
        self.tasks.values().collect()
    }
    
    pub fn get_pending_tasks(&self) -> Vec<&Task> {
        self.tasks.values()
            .filter(|t| !t.completed)
            .collect()
    }
    
    pub fn get_completed_tasks(&self) -> Vec<&Task> {
        self.tasks.values()
            .filter(|t| t.completed)
            .collect()
    }
    
    pub fn get_tasks_by_priority(&self, priority: &str) -> Vec<&Task> {
        self.tasks.values()
            .filter(|t| t.priority == priority)
            .collect()
    }
    
    pub fn task_count(&self) -> usize {
        self.tasks.len()
    }
    
    pub fn completed_count(&self) -> usize {
        self.tasks.values()
            .filter(|t| t.completed)
            .count()
    }
}

impl Default for TaskManager {
    fn default() -> Self {
        Self::new()
    }
}
